ecs_pod_creator.go

```diff
@@ -68,18 +88,81 @@ func (o *ECSPodCreationOptions) SetCPU(cpu int) *ECSPodCreationOptions {
 	return o
 }
 
-// SetTags sets the tags for the pod. This overwrites any existing tags.
+// SetTaskRole sets the task role that the pod can use.
+func (o *ECSPodCreationOptions) SetTaskRole(role string) *ECSPodCreationOptions {
+	o.TaskRole = &role
+	return o
+}
+
+// SetTags sets the tags for the pod definition. This overwrites any existing
+// tags.
 func (o *ECSPodCreationOptions) SetTags(tags []string) *ECSPodCreationOptions {
 	o.Tags = tags
 	return o
 }
 
-// AddTags adds new tags to the existing ones for the pod.
+// AddTags adds new tags to the existing ones for the pod definition.
 func (o *ECSPodCreationOptions) AddTags(tags ...string) *ECSPodCreationOptions {
 	o.Tags = append(o.Tags, tags...)
 	return o
 }
 
+// SetExecutionOptions sets the options to configure how the pod executes.
+func (o *ECSPodCreationOptions) SetExecutionOptions(opts ECSPodExecutionOptions) *ECSPodCreationOptions {
+	o.ExecutionOpts = &opts
+	return o
+}
+
+// Validate checks that all the required parameters are given and the values are
+// valid.
+func (o *ECSPodCreationOptions) Validate() error {
+	catcher := grip.NewBasicCatcher()
+	catcher.NewWhen(o.Name != nil && *o.Name == "", "cannot specify an empty name")
+	catcher.NewWhen(o.MemoryMB != nil && *o.MemoryMB <= 0, "must have positive memory value if non-default")
+	catcher.NewWhen(o.CPU != nil && *o.CPU <= 0, "must have positive CPU value if non-default")
+	catcher.NewWhen(len(o.ContainerDefinitions) == 0, "must specify at least one container definition")
+	var totalContainerMemMB, totalContainerCPU int
+	for _, def := range o.ContainerDefinitions {
+		catcher.Wrapf(def.Validate(), "container definition '%s'", def.Name)
+		if def.MemoryMB != nil {
+			totalContainerMemMB += *def.MemoryMB
+		} else if o.MemoryMB == nil {
+			catcher.Errorf("must specify container-level memory to allocate for each container if pod-level memory is not specified")
+		}
+
+		if def.CPU != nil {
+			totalContainerCPU += *def.CPU
+		} else if o.CPU == nil {
+			catcher.Errorf("must specify container-level CPU to allocate for each container if pod-level CPU is not specified")
+		}
+	}
+	if o.MemoryMB != nil {
```

Comment:
If `MemoryMB` and `CPU` are not specified, do we have a default value?

Reply:
Also, if they are specified and the analogous container options are not, how do we allocate memory and CPU across containers? Do we just divide evenly?

Reply:
If memory is unspecified at the pod level, you have to give it for all the containers within the pod. Same with CPU. If they're only set at the pod level, all the containers share the memory/CPU pool up to the pod-level memory/CPU limit and contend for the resources on a first-come-first-serve basis. For example, if one container is using up all the memory available and another one in the pod asks for more memory, it'll be killed by the OOM killer.

This is mostly because ECS has to know how much memory/CPU to allocate to it when placing it on EC2 instances.